zone/expedition.cpp

```diff
@@ -0,0 +1,2275 @@
+	ExpeditionDatabase::AddLockoutDuration(m_members, lockout, modified_seconds);
+
+	ProcessLockoutDuration(lockout, seconds, members_only);
+	SendWorldLockoutDuration(lockout, seconds, members_only);
+}
+
+void Expedition::UpdateLockoutDuration(
+	const std::string& event_name, uint32_t seconds, bool members_only)
+{
+	// some live expeditions update existing lockout timers during progression
+	auto it = m_lockouts.find(event_name);
+	if (it != m_lockouts.end())
+	{
+		seconds = static_cast<uint32_t>(seconds * RuleR(Expedition, LockoutDurationMultiplier));
+
+		uint64_t expire_time = it->second.GetStartTime() + seconds;
+		AddLockout({ m_uuid, m_expedition_name, event_name, expire_time, seconds }, members_only);
+	}
+}
+
+void Expedition::RemoveLockout(const std::string& event_name)
+{
+	ExpeditionDatabase::DeleteLockout(m_id, event_name);
+	ExpeditionDatabase::DeleteMembersLockout(m_members, m_expedition_name, event_name);
+
+	ExpeditionLockoutTimer lockout{m_uuid, m_expedition_name, event_name, 0, 0};
+	ProcessLockoutUpdate(lockout, true);
+	SendWorldLockoutUpdate(lockout, true);
+}
+
+void Expedition::AddInternalMember(
+	const std::string& char_name, uint32_t character_id, ExpeditionMemberStatus status)
+{
+	auto it = std::find_if(m_members.begin(), m_members.end(),
+		[character_id](const ExpeditionMember& member) {
+			return member.char_id == character_id;
+		});
+
+	if (it == m_members.end())
+	{
+		m_members.emplace_back(character_id, char_name, status);
+	}
+}
+
+bool Expedition::AddMember(const std::string& add_char_name, uint32_t add_char_id)
+{
+	if (HasMember(add_char_id))
+	{
+		return false;
+	}
+
+	ExpeditionDatabase::InsertMember(m_id, add_char_id);
+	m_dynamiczone.AddCharacter(add_char_id);
+
+	ProcessMemberAdded(add_char_name, add_char_id);
+	SendWorldMemberChanged(add_char_name, add_char_id, false);
+
+	return true;
+}
+
+void Expedition::RemoveAllMembers(bool enable_removal_timers)
+{
+	m_dynamiczone.RemoveAllCharacters(enable_removal_timers);
+
+	ExpeditionDatabase::DeleteAllMembersPendingLockouts(m_members);
+	ExpeditionDatabase::DeleteAllMembers(m_id);
+
+	SendUpdatesToZoneMembers(true);
+	SendWorldExpeditionUpdate(ServerOP_ExpeditionMembersRemoved);
+
+	m_members.clear();
+}
+
+bool Expedition::RemoveMember(const std::string& remove_char_name)
+{
+	auto member = GetMemberData(remove_char_name);
+	if (member.char_id == 0 || member.name.empty())
+	{
+		return false;
+	}
+
+	ExpeditionDatabase::DeleteMember(m_id, member.char_id);
+	m_dynamiczone.RemoveCharacter(member.char_id);
+
+	ProcessMemberRemoved(member.name, member.char_id);
+	SendWorldMemberChanged(member.name, member.char_id, true);
+
+	// live always sends a leader update but we can send only if leader changes
+	if (member.char_id == m_leader.char_id)
+	{
+		ChooseNewLeader();
+	}
+
+	return true;
+}
+
+void Expedition::SwapMember(Client* add_client, const std::string& remove_char_name)
```

Comment:
What exactly is this swapping?

Reply:
This is for the client's /dzswap command which takes <add name> <remove name> arguments. The actual swap only occurs when the added player accepts the invite so add name is the add client in this case